cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -2141,28 +2141,39 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
         }
 
         Opcode::FcvtFromSint => {
-            let (ext_spec, src_size) = match ctx.input_ty(insn, 0) {
-                types::I8 | types::I16 => (Some(ExtSpec::SignExtendTo32), OperandSize::Size32),
-                types::I32 => (None, OperandSize::Size32),
-                types::I64 => (None, OperandSize::Size64),
-                _ => unreachable!(),
-            };
+            let output_ty = ty.unwrap();
+            if !output_ty.is_vector() {
+                let (ext_spec, src_size) = match ctx.input_ty(insn, 0) {
+                    types::I8 | types::I16 => (Some(ExtSpec::SignExtendTo32), OperandSize::Size32),
+                    types::I32 => (None, OperandSize::Size32),
+                    types::I64 => (None, OperandSize::Size64),
+                    _ => unreachable!(),
+                };
 
-            let src = match ext_spec {
-                Some(ext_spec) => RegMem::reg(extend_input_to_reg(ctx, inputs[0], ext_spec)),
-                None => input_to_reg_mem(ctx, inputs[0]),
-            };
+                let src = match ext_spec {
+                    Some(ext_spec) => RegMem::reg(extend_input_to_reg(ctx, inputs[0], ext_spec)),
+                    None => input_to_reg_mem(ctx, inputs[0]),
+                };
 
-            let output_ty = ty.unwrap();
-            let opcode = if output_ty == types::F32 {
-                SseOpcode::Cvtsi2ss
+                let opcode = if output_ty == types::F32 {
+                    SseOpcode::Cvtsi2ss
+                } else {
+                    assert_eq!(output_ty, types::F64);
+                    SseOpcode::Cvtsi2sd
+                };
+                let dst = get_output_reg(ctx, outputs[0]);
+                ctx.emit(Inst::gpr_to_xmm(opcode, src, src_size, dst));
             } else {
-                assert_eq!(output_ty, types::F64);
-                SseOpcode::Cvtsi2sd
-            };
-
-            let dst = get_output_reg(ctx, outputs[0]);
-            ctx.emit(Inst::gpr_to_xmm(opcode, src, src_size, dst));
+                let ty = ty.unwrap();
+                let src = put_input_in_reg(ctx, inputs[0]);
+                let dst = get_output_reg(ctx, outputs[0]);
+                let opcode = match ctx.input_ty(insn, 0) {
+                    types::I32X4 => SseOpcode::Cvtdq2ps,
+                    _ => unreachable!("Vector type for conversion"),
```

Comment:
```suggestion
                    _ => unimplemented!("unable to use type {} for op {}", ctx.input_ty(insn, 0), op),
```
My reasoning: more information is usually better with these errors and [fcvt_from_sint](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/codegen/meta/src/shared/instructions.rs#L4231) can accept more than the `I32X4` type (so other types are unimplemented, not unreachable).

Reply:
@abrown Yes .. agree to more information. The unreachable vs unimplemented I um .. kind of have a different take. Right now, until cliff instructions are added, it is unreachable (I think?). There is no clif instruction (I think?) that can get here .. as this code is in a branch that has already checked for a vector type. Taken in context, the unreachable doesn't refer to just  fcvt_from_sint it refers to vector types of fcvt_from_sint in which case there is only one afaik? I will change it though because taken out of context it is reachable and in any case I think with this kind of ambiguity, for convenience power goes to the who reviews  :-) .. and of course precedence/consistency. 👍

Reply:
> it refers to vector types of fcvt_from_sint in which case there is only one afaik

Take a look at the types of the fcvt_from_sint instruction I linked to: `x` is `Int`, which makes that CLIF instruction polymorphic over all integer and lane sizes. So someone could theoretically write a CLIF instruction with a type that would trigger the error, even if the Wasm translation would never create it.